sdk/lib/apisets/apisets.c

```diff
@@ -0,0 +1,100 @@
+ * COPYRIGHT:   Copyright 2024 Mark Jansen <mark.jansen@reactos.org>
+ */
+
+#include <ndk/umtypes.h>
+#include <ndk/rtlfuncs.h>
+#include "apisetsp.h"
+
+
+const ULONGLONG API_ = (ULONGLONG)0x2D004900500041; /// L"API-"
+const ULONGLONG EXT_ = (ULONGLONG)0x2D005400580045; /// L"EXT-";
+
+WORD PrefixSize = sizeof(L"api-") - sizeof(WCHAR);
+WORD ExtensionSize = sizeof(L".dll") - sizeof(WCHAR);
+
+
+// The unittest wants to link against ntdll,
+// but the apisets library is used _inside_ ntdll so we have to make the symbol work in both cases.
+// Therefore, we ask the callers of this library to define this symbol (and initialize it with RtlCompareUnicodeString)
+extern LONG (NTAPI *
+ApisetpRtlCompareUnicodeString)(
+    IN PCUNICODE_STRING String1,
+    IN PCUNICODE_STRING String2,
+    IN BOOLEAN CaseInSensitive);
+
+
+// The official prototype according to the Windows kit 8.1 is:
+//NTSTATUS
+//ApiSetResolveToHost (
+//    _In_ PCAPI_SET_NAMESPACE_ARRAY Schema,
+//    _In_ PCUNICODE_STRING FileNameIn,
+//    _In_opt_ PCUNICODE_STRING ParentName,
+//    _Out_ PBOOLEAN Resolved,
+//    _Out_ PUNICODE_STRING HostBinary
+//    );
+
+
+NTSTATUS
+ApiSetResolveToHost(
+    _In_ DWORD ApisetVersion,
+    _In_ PCUNICODE_STRING ApiToResolve,
+    _Out_ PBOOLEAN Resolved,
+    _Out_ PUNICODE_STRING Output
+)
+{
+    if (ApiToResolve->Length < PrefixSize)
+    {
+        *Resolved = FALSE;
+        return STATUS_SUCCESS;
+    }
+
+    // Grab the first four chars from the string, converting the first 3 to uppercase
+    PWSTR ApiSetNameBuffer = ApiToResolve->Buffer;
+    ULONGLONG ApiSetNameBufferPrefix = ((ULONGLONG *)ApiSetNameBuffer)[0] & 0xFFFFFFDFFFDFFFDF;
+    // Check if it matches either 'api-' or 'ext-'
+    if (!(ApiSetNameBufferPrefix == API_ || ApiSetNameBufferPrefix == EXT_))
+    {
+        *Resolved = FALSE;
+        return STATUS_SUCCESS;
+    }
+
+    // If there is an extension, cut it off (we store apisets without extension)
+    UNICODE_STRING Tmp = *ApiToResolve;
+    if (!_wcsnicmp(Tmp.Buffer + (Tmp.Length - ExtensionSize) / sizeof(WCHAR), L".dll", ExtensionSize / sizeof(WCHAR)))
+        Tmp.Length -= ExtensionSize;
+
+    // Binary search the apisets
+    LONG UBnd = g_ApisetsCount - 1;
+    LONG LBnd = 0;
+    while (LBnd <= UBnd)
+    {
+        LONG Index = (UBnd - LBnd) / 2 + LBnd;
+
+        LONG result = ApisetpRtlCompareUnicodeString(&Tmp, &g_Apisets[Index].Name, TRUE);
+        if (result == 0)
+        {
+            // Check if this version is included
+            if (g_Apisets[Index].dwOsVersions & ApisetVersion)
+            {
+                // Return a static string (does not have to be freed)
+                *Resolved = TRUE;
+                *Output = g_Apisets[Index].Target;
+            }
+            return STATUS_SUCCESS;
+        }
+        else if (result < 0)
+        {
+            UBnd = Index - 1;
+        }
+        else
+        {
+            LBnd = Index + 1;
+        }
+    }
```

Comment:
```suggestion
    ApiSet = bsearch((PCHAR)&Tmp,
                     (PCHAR)&g_Apisets,
                     ARRAYSIZE(g_Apisets),
                     sizeof(g_Apisets[0]),
                     ApisetCompareFunc);
    if (ApiSet != NULL)
    {
            // Check if this version is included
            if (g_Apisets[Index].dwOsVersions & ApisetVersion)
            {
                // Return a static string (does not have to be freed)
                *Resolved = TRUE;
                *Output = g_Apisets[Index].Target;
            }
            return STATUS_SUCCESS;
        }
```

Reply:
Yeah, bsearch might make sense now.
I usually go for the inline written one, because that gives you the insertion point if it was not found, but we don't need that here.

Reply:
That drags in another dependency which further complicates the whole `rtl_um` / `rtl_unittest` dance.